File: shared-components/src/commonMain/kotlin/dev/viethung/components/numberinput/NumberInputViewModel.kt
package dev.viethung.components.numberinput

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class NumberInputViewModel(
    private val formatter: LocaleNumberFormatter,
    initialValue: Double? = null,
    private val significantDigits: Int = 2,
    private val locale: String = "en-US",
    private val allowNegative: Boolean = true,
) : ViewModel() {

    // Per Validation Summary decision #2 / AC-08: clamp negative initialValue to 0.0 when allowNegative=false.
    private val seedValue: Double? =
        if (!allowNegative && initialValue != null && initialValue < 0.0) 0.0 else initialValue

    private val _state: MutableStateFlow<NumberInputUiState> = MutableStateFlow(
        NumberInputUiState.Idle(
            rawText = seedValue?.toString().orEmpty(),
            formattedText = seedValue?.let { formatter.format(it, significantDigits, locale) }.orEmpty(),
            value = seedValue,
            significantDigits = significantDigits,
            locale = locale,
            allowNegative = allowNegative,
        )
    )
    val state: StateFlow<NumberInputUiState> = _state.asStateFlow()

    fun onFocusChanged(focused: Boolean) {
        viewModelScope.launch {
            val current = _state.value
            if (focused) {
                if (current !is NumberInputUiState.Editing) {
                    // Enter Editing with the locale-correct formatted form as both rawText and
                    // formattedText, so the user starts editing the grouped string they see.
                    val carry = current.formattedText
                    _state.value = NumberInputUiState.Editing(
                        rawText = carry,
                        formattedText = carry,
                        value = current.value,
                        significantDigits = significantDigits,
                        locale = locale,
                        allowNegative = allowNegative,
                    )
                }
            } else {
                commitAndSetIdle()
            }
        }
    }

    fun onTextChange(newRawText: String) {
        viewModelScope.launch {
            val current = _state.value
            val parsed = formatter.parse(newRawText, locale)
            // Per implementation notes §4.2:
            // - If rawText is blank → value is null (empty field).
            // - If parse fails on non-blank rawText: keep prior value, update rawText only.
            // - Per Validation Summary decision #2: if allowNegative=false and parsed is negative,
            //   reject the value change (keep prior value), but update rawText.
            val newValue = when {
                newRawText.isBlank() -> null
                parsed == null -> current.value
                !allowNegative && parsed < 0.0 -> current.value
                else -> parsed
            }
            _state.value = NumberInputUiState.Editing(
                rawText = newRawText,
                formattedText = formatter.formatLive(newRawText, locale),
                value = newValue,
                significantDigits = significantDigits,
                locale = locale,
                allowNegative = allowNegative,
            )
        }
    }

    fun onToggleSign() {
        // No-op if null value or allowNegative=false
        if (!allowNegative) return
        viewModelScope.launch {
            val current = _state.value
            val currentValue = current.value ?: return@launch
            val toggled = -currentValue
            // Use locale-aware formatted form as both rawText and formattedText so vi-VN/de-DE
            // get the right decimal separator (Double.toString always uses '.').
            val newRawText = formatter.format(toggled, significantDigits, locale)
            _state.value = NumberInputUiState.Editing(
                rawText = newRawText,
                formattedText = newRawText,
                value = toggled,
                significantDigits = significantDigits,
                locale = locale,
                allowNegative = allowNegative,
            )
        }
    }

    fun onClear() {
        viewModelScope.launch {
            _state.value = NumberInputUiState.Editing(
                rawText = "",
                formattedText = "",
                value = null,
                significantDigits = significantDigits,
                locale = locale,
                allowNegative = allowNegative,
            )
        }
    }

    fun onCommit() {
        viewModelScope.launch {
            commitAndSetIdle()
        }
    }

    private suspend fun commitAndSetIdle() {
        val current = _state.value
        val formatted = current.value?.let { formatter.format(it, significantDigits, locale) }.orEmpty()
        val committed = NumberInputUiState.Committed(
            rawText = current.rawText,
            formattedText = formatted,
            value = current.value,
            significantDigits = significantDigits,
            locale = locale,
            allowNegative = allowNegative,
        )
        _state.value = committed
        _state.value = NumberInputUiState.Idle(
            rawText = committed.rawText,
            formattedText = committed.formattedText,
            value = committed.value,
            significantDigits = significantDigits,
            locale = locale,
            allowNegative = allowNegative,
        )
    }
}

File: swift-package/NumberInputKit/Sources/NumberInputKit/NumberInputViewModel.swift
import Foundation
import Combine

/// Pure-Swift port of the Kotlin `NumberInputViewModel`.
/// Owns all state as `@Published var state: NumberInputUIState`. Methods are synchronous —
/// no coroutines. State transitions are identical to the Kotlin VM.
public final class NumberInputViewModel: ObservableObject {

    @Published public private(set) var state: NumberInputUIState

    private let formatter: any LocaleNumberFormatter
    private let significantDigits: Int
    private let locale: String
    private let allowNegative: Bool

    public init(
        formatter: any LocaleNumberFormatter = newLocaleNumberFormatter(),
        initialValue: Double? = nil,
        config: NumberInputConfig = NumberInputConfig()
    ) {
        self.formatter = formatter
        self.significantDigits = config.significantDigits
        self.locale = config.locale
        self.allowNegative = config.allowNegative

        // AC-08: clamp negative initialValue to 0.0 when allowNegative=false
        let seedValue: Double? = {
            if !config.allowNegative, let v = initialValue, v < 0.0 { return 0.0 }
            return initialValue
        }()

        let formattedText = seedValue.map { formatter.format($0, significantDigits: config.significantDigits, locale: config.locale) } ?? ""
        let rawText = formattedText
        self.state = .idle(NumberInputUIState.Payload(
            rawText: rawText,
            formattedText: formattedText,
            value: seedValue,
            significantDigits: config.significantDigits,
            locale: config.locale,
            allowNegative: config.allowNegative
        ))
    }

    /// Convenience init matching the Kotlin VM signature (individual parameters).
    public init(
        formatter: any LocaleNumberFormatter = newLocaleNumberFormatter(),
        initialValue: Double? = nil,
        significantDigits: Int = 2,
        locale: String = "en-US",
        allowNegative: Bool = true
    ) {
        self.formatter = formatter
        self.significantDigits = significantDigits
        self.locale = locale
        self.allowNegative = allowNegative

        let seedValue: Double? = {
            if !allowNegative, let v = initialValue, v < 0.0 { return 0.0 }
            return initialValue
        }()

        let formattedText = seedValue.map { formatter.format($0, significantDigits: significantDigits, locale: locale) } ?? ""
        self.state = .idle(NumberInputUIState.Payload(
            rawText: formattedText,
            formattedText: formattedText,
            value: seedValue,
            significantDigits: significantDigits,
            locale: locale,
            allowNegative: allowNegative
        ))
    }

    public func onFocusChanged(focused: Bool) {
        if focused {
            guard case .editing = state else {
                // Enter Editing, carrying the locale-correct formatted form
                let carry = state.payload.formattedText
                state = .editing(payload(rawText: carry, formattedText: carry, value: state.payload.value))
                return
            }
        } else {
            commitAndSetIdle()
        }
    }

    public func onTextChange(_ newRawText: String) {
        let current = state.payload
        let parsed = formatter.parse(newRawText, locale: locale)
        let newValue: Double?
        if newRawText.isEmpty || newRawText == "-" {
            newValue = newRawText.isEmpty ? nil : current.value
        } else if let p = parsed {
            newValue = (!allowNegative && p < 0.0) ? current.value : p
        } else {
            // parse failed on non-blank: keep prior value
            newValue = current.value
        }
        let formattedText = formatter.formatLive(newRawText, locale: locale)
        state = .editing(payload(rawText: newRawText, formattedText: formattedText, value: newValue))
    }

    public func onToggleSign() {
        guard allowNegative else { return }
        guard let currentValue = state.payload.value else { return }
        let toggled = -currentValue
        let newRawText = formatter.format(toggled, significantDigits: significantDigits, locale: locale)
        state = .editing(payload(rawText: newRawText, formattedText: newRawText, value: toggled))
    }

    public func onClear() {
        state = .editing(payload(rawText: "", formattedText: "", value: nil))
    }

    public func onCommit() {
        commitAndSetIdle()
    }

    // MARK: - Private

    private func commitAndSetIdle() {
        let current = state.payload
        let formatted = current.value.map { formatter.format($0, significantDigits: significantDigits, locale: locale) } ?? ""
        let committedPayload = NumberInputUIState.Payload(
            rawText: current.rawText,
            formattedText: formatted,
            value: current.value,
            significantDigits: significantDigits,
            locale: locale,
            allowNegative: allowNegative
        )
        // Emit Committed, then immediately Idle — matches Kotlin VM ordering.
        state = .committed(committedPayload)
        state = .idle(committedPayload)
    }

    private func payload(rawText: String, formattedText: String, value: Double?) -> NumberInputUIState.Payload {
        NumberInputUIState.Payload(
            rawText: rawText,
            formattedText: formattedText,
            value: value,
            significantDigits: significantDigits,
            locale: locale,
            allowNegative: allowNegative
        )
    }
}
